Support connecting to BMP devices via TCP

xobs opened the pull request:
## Detailed description

TCP connections can behave similarly to sockets on all supported platforms. This patch allows for specifying "address:port" on Linux, Windows, and Mac for connecting to a remote device.

Currently this has been tested with Farpatch, however future versions could allow inception debugging wherein low-level commands are forwarded from BMDA to the target, which could even be a second instance of BMDA.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (see [Building the firmware](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-firmware))
* [x] It builds as BMDA (see [Building the BMDA](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-app))
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -43,6 +47,73 @@ static uint8_t read_buffer[READ_BUFFER_LENGTH];
+	};
+	struct addrinfo *server_info;
+
+	if (!name) {
+		return false;
+	}
```
The braces on this are superfluous and can be dropped. This check should ideally be done at the top of the function - we are in C11, so there is no requirement that variables are defined first (it actually benefits stack usage and such for them to be declared as needed). Please include a length check on `name` to warn users of truncation given the strncpy() into a stack buffer.

xobs replied:
My Rust is showing here.

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -43,6 +47,73 @@ static uint8_t read_buffer[READ_BUFFER_LENGTH];
+
+	service_name = strstr(hostname, ":");
+	if (service_name == NULL) {
+		return false;
```
It might be helpful to `DEBUG_WARN` why this code is ignoring the name string as not-a-`hostname:port`.

xobs replied:
That will get printed out as part of every invocation. Is that okay? I thought this might be a niche case so I wanted to avoid reporting anything like that.

dragonmux replied:
Given that should only happen if BMDA was unable to open the target as a file/device and it's then failing as a network socket target - that seems completely reasonable as it helps the user diagnose what's going on.

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -43,6 +47,73 @@ static uint8_t read_buffer[READ_BUFFER_LENGTH];
+
+	// Separate the service name / port number from the hostname
+	*service_name = '\0';
+	service_name += 1;
```
`++service_name;`? Seeing this only adds 1 to the pointer.

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -43,6 +47,73 @@ static uint8_t read_buffer[READ_BUFFER_LENGTH];
+		}
+
+		if (connect(fd, p->ai_addr, p->ai_addrlen) == -1) {
+			close(fd);
```
So this code is semantically correct if ported to Windows, please use `closesocket()` here and use something like this just before the `try_opening_network_device()` function definition:
```c
#ifndef _WIN32
inline int closesocket(const int socket) { return close(socket); }
#endif
```

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -43,6 +47,73 @@ static uint8_t read_buffer[READ_BUFFER_LENGTH];
+	}
+
+	// Loop through all the results and connect to the first we can.
+	struct addrinfo *p;
+	for (p = server_info; p != NULL; p = p->ai_next) {
```
So we don't wind up introducing an unexplained single letter variable name, we're going to suggest that `server_info` be renamed to `results` and `p` to `address` or `server_addr` or similar.

dragonmux reviewed (commented):
We've primarily reviewed the Unix implementation of this as most of the same comments apply straight to the Windows one too. With the review notes taken care of (noting the extra braces one applies in a lot of places), we'll review the Windows code properly and more closely. Apologies this review's taken so long to get to.

xobs commented:
I've rebased this on top of the latest `main`, and made the requested changes. I also updated `serial_win.c` with the same fixes such as renaming variables and adding `DEBUG_WARN()` in various places.

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -43,6 +47,84 @@ static uint8_t read_buffer[READ_BUFFER_LENGTH];
+	*service_name = '\0';
+	++service_name;
+
+	if (*service_name == '\0') {
+		return false;
+	}
+
+	if (getaddrinfo(hostname, service_name, &addr_hints, &results) != 0) {
+		return false;
+	}
```
These and the other single-statement `if` blocks below can all do with their parens dropping please.

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -43,6 +47,84 @@ static uint8_t read_buffer[READ_BUFFER_LENGTH];
+	char hostname[256];
+
+	// The service name or port number
+	char *service_name;
```
Please move the definition of this variable down to where it gets initialised on line 85.

dragonmux commented on src/platforms/hosted/serial_win.c:
```diff
@@ -41,11 +43,97 @@
+
+	// The service name is commonly empty on Windows where port names
+	// tend to be things like "COM4:" which can be mistaken for a hostname.
+	if (*service_name == '\0') {
+		return false;
+	}
```
Same comment about single-statement `if`'s.

dragonmux commented on src/platforms/hosted/serial_win.c:
```diff
@@ -256,8 +347,15 @@ bool serial_open(const bmda_cli_options_s *const cl_opts, const char *const seri
+	if (port_handle != INVALID_HANDLE_VALUE) {
+		CloseHandle(port_handle);
+		port_handle = INVALID_HANDLE_VALUE;
+		WSACleanup();
```
Does this call not need to happen in the network `if` rather than this one?

xobs replied:
Oof, that's a really good catch.

dragonmux commented on src/platforms/hosted/serial_win.c:
```diff
@@ -266,16 +364,66 @@ bool platform_buffer_write(const void *const data, const size_t length)
 	return true;
 }
 
+static ssize_t bmda_read_more_socket_data()
```
This needs its parens set to `(void)`. While this compiles it [accidentally invokes the non-prototype function declaration](https://en.cppreference.com/w/c/language/function_declaration#Syntax) syntax which has undesirable semantics in this context.

dragonmux commented on src/platforms/hosted/serial_win.c:
```diff
@@ -266,16 +364,66 @@ bool platform_buffer_write(const void *const data, const size_t length)
+
 static ssize_t bmda_read_more_data(const uint32_t end_time)
 {
+	if (network_socket != INVALID_SOCKET) {
+		return bmda_read_more_socket_data();
+	}
```
Same comment about single-statement `if`'s.

dragonmux reviewed (commented):
There are a couple of queries/niggles in the Windows code, with those sorted and some straggling single-statement `if` with braces instances sorted, this is looking good to merge.

xobs commented:
I've made those corrections. Well spotted.

dragonmux commented on src/platforms/hosted/serial_win.c:
```diff
@@ -202,6 +286,9 @@ bool serial_open(const bmda_cli_options_s *const cl_opts, const char *const seri
 
 	/* If opening the device node failed for any reason, error out early */
 	if (port_handle == INVALID_HANDLE_VALUE) {
+		if (try_opening_network_device(cl_opts->opt_device)) {
+			return true;
+		}
```
Here's another instance of the signle-statement `if`.

dragonmux commented on src/platforms/hosted/serial_unix.c:
```diff
@@ -97,6 +173,9 @@ bool serial_open(const bmda_cli_options_s *cl_opts, const char *serial)
 	}
 	fd = open(name, O_RDWR | O_SYNC | O_NOCTTY);
 	if (fd < 0) {
+		if (try_opening_network_device(name)) {
+			return true;
+		}
```
Same comment about single-statement `if`'s, and likewise for the one below this.

dragonmux reviewed (commented):
There seem to be just 3 straggler instances of the single-statement `if` braces lint, and with those solved, this is ready.

dragonmux reviewed (approved):
This all LGTM now, merging once the builds complete. Thank you very much for the contribution!
